pkg/providers/capacityreservation/provider.go

```diff
@@ -0,0 +1,102 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package capacityreservation
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	sdk "github.com/aws/karpenter-provider-aws/pkg/aws"
+)
+
+type Provider interface {
+	List(context.Context, ...v1.CapacityReservationSelectorTerm) ([]*ec2types.CapacityReservation, error)
+}
+
+type DefaultProvider struct {
+	sync.RWMutex
+
+	ec2api sdk.EC2API
+	clk    clock.Clock
+	cache  *cache.Cache
+	cm     *pretty.ChangeMonitor
+}
+
+func NewProvider(ec2api sdk.EC2API, clk clock.Clock, cache *cache.Cache) *DefaultProvider {
+	return &DefaultProvider{
+		ec2api: ec2api,
+		clk:    clk,
+		cache:  cache,
+		cm:     pretty.NewChangeMonitor(),
+	}
+}
+
+func (p *DefaultProvider) List(ctx context.Context, selectorTerms ...v1.CapacityReservationSelectorTerm) ([]*ec2types.CapacityReservation, error) {
+	queries := QueriesFromSelectorTerms(selectorTerms...)
+	reservations, remainingQueries := func() ([]*ec2types.CapacityReservation, []*Query) {
+		p.RLock()
+		defer p.RUnlock()
+		reservations := []*ec2types.CapacityReservation{}
+		remaining := []*Query{}
+		for _, query := range queries {
+			if value, ok := p.cache.Get(query.CacheKey()); ok {
+				reservations = append(reservations, value.([]*ec2types.CapacityReservation)...)
+			} else {
+				remaining = append(remaining, query)
+			}
+		}
+		return reservations, remaining
+	}()
+	if len(remainingQueries) == 0 {
+		return p.filterReservations(reservations), nil
+	}
+
+	p.Lock()
```

Comment:
Same question here -- isn't this thread-safe? I thought the main point of these locks was that, if there were parallel calls, that they would wait until all of the direct reads were resolved so that the following calls would all hit the cache